docs/reference/transforms/task_context.rst

```diff
@@ -0,0 +1,135 @@
+.. _task_context:
+
+Task Context
+============
+
+The :mod:`taskgraph.transforms.task_context` transform can be used to
+substitute values into any field in a task with data that is not known
+until ``taskgraph`` runs.
+
+This data can be provided in a few ways, as described below.
+
+Usage
+-----
+
+from-parameters
+~~~~~~~~~~~~~~~
+
+Context can be pulled from parameters that are provided to ``taskgraph``.
+
+First, add the transform to the ``transforms`` key in your ``kind.yml`` file:
+
+.. code-block:: yaml
+
+   transforms:
+     - taskgraph.transforms.task_context
+     # ...
+
+Then create a ``task-context`` section in your task definition, e.g:
+
+.. code-block:: yaml
+
+   tasks:
+     build:
+       description: my description {foo}
+       task-context:
+         from-parameters:
+           foo: foo
+         substitution-fields:
+           - description
+
+When ``taskgraph`` is run the value of ``foo`` in the parameters it is provided
+will be substituted into the description. For example, if the following parameters
+are used:
+
+.. code-block:: yaml
+
+   foo: with some extra
+
+...the description will end up with a value of "my description with some extra".
+
+When using ``from-parameters`` you may also provide an ordered list of keys to
+look for in the parameters, with the first one found being used. For example,
+with the this ``kind``:
+
+.. code-block:: yaml
+
+   tasks:
+     build:
+       description: my description {foo}
+       task-context:
+         from-parameters:
+           foo:
+             - foo
+             - default
+         substitution-fields:
+           - description
+
+...the description will bring in the value ``foo`` from the parameters if
+present, or ``default`` otherwise.
+
+from-file
+~~~~~~~~~
+
+Context may also be provided from a defined yaml file. The provided file
+should usually only contain top level keys and values (eg: nested objects
+will not be interpolated - they will be substituted as text representations
+of the object).
+
+For example, with this kind definition:
+
+.. code-block:: yaml
+
+   tasks:
+     build:
+       description: my description {foo}
+       task-context:
+         from-file: some_file.yaml
+         substitution-fields:
+           - description
+
+And this in ``some_file.yaml``:
+
+.. code-block:: yaml
+
+   foo: from a file
+
+...description will end up with "my description from a file".
+
+
+Direct Context
+~~~~~~~~~~~~~~
+
+You may also specify context as direct keys and values in the ``task-context``
+configuration. This can be useful in ``kinds`` that define most of their
+contents in ``task-defaults``, but have some values that may differ for
+various concrete ``tasks`` in the ``kind``.
+
+For example:
+
+.. code-block:: yaml
+
+   task-defaults:
+     description: my description {extra_desc}
+     task-context:
+       substitution-fields:
+         - description
+
+   tasks:
+     build1:
+       task-context:
+         extra_desc: build1
```

Comment:
I'm confused why the direct setting is needed, wouldn't this also work if they had:
```
task-context:
    from-parameters:
        extra_desc: build1
```

The `task-defaults` use a merge deep function, so it should work I think. But maybe I'm missing something.

Reply:
This is most useful when you have multiple entries in a `kind` that are mostly the same, but whose specific `task` entries may need different values in, eg: a command. https://github.com/mozilla/firefox-translations-training/blob/main/taskcluster/ci/merge-mono/kind.yml is a good example of this (it's still using `command-context`, but I think you'll get the point). These tasks are nearly identical, except that one uses `max_sent` that comes from `training_config.experiment.mono-max-sentences-src` in parameters, while the other uses `max_sent` from `training_config.experiment.mono-max-sentences-trg`.

Reply:
Yeah sorry, I was confused because I thought `from-parameters` was arbitrary key/values as well and I didn't understand why you needed to have two methods of setting arbitrary key/values.

What if instead of setting these at the root of the `task-context` dict, we put them under a `from-object` or `from-dict` key instead? That way we don't pollute the top namespace with arbitrary keys and it's more consistent with the other two methods of providing context.

Reply:
That's an excellent idea - will do.